# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [1]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [2]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [3]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [4]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [5]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [6]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [7]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [8]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [9]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [10]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [11]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [12]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [13]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [14]:
df['Season'].nunique()

32

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [15]:
df['Wteam'].value_counts()

Wteam
1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: count, Length: 364, dtype: int64

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [16]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [17]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [18]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [19]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [20]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [21]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [22]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [23]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [24]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [25]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [26]:
df.sort_values(by='Season')
df.sort_values(by='Wteam')
df.sort_values(by='Wscore')
df.iloc[100:201, [0, 1, 2, 3]]

Season  Daynum  Wteam  Wscore
100    1985      29   1409      86
101    1985      29   1444      90
102    1985      29   1451      89
103    1985      30   1139      78
104    1985      30   1154      78
..      ...     ...    ...     ...
196    1985      33   1203      83
197    1985      33   1222      73
198    1985      33   1227      76
199    1985      33   1228      73
200    1985      33   1229      54

[101 rows x 4 columns]

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [27]:
index_of_max_winning_score=df['Wscore'].idxmax()

In [28]:
season_highest_score=df.loc[index_of_max_winning_score,'Season']
winning_team_highest_score=df.loc[index_of_max_winning_score,'Wteam']

In [29]:
print("Season and Winning Team for the Game with the Highest Winning Score:")
print("Season:", season_highest_score)
print("Wteam:", winning_team_highest_score)



Season and Winning Team for the Game with the Highest Winning Score:
Season: 1991
Wteam: 1258


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [30]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [31]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [32]:
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

In [33]:
average_score_difference = df['ScoreDifference'].mean()

In [34]:
filtered_df = df[df['ScoreDifference'] > average_score_difference]

In [35]:
print(filtered_df.head())

    Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference
0     1985      20   1228      81   1328      64    N      0               17
3     1985      25   1165      70   1432      54    H      0               16
6     1985      25   1228      64   1226      44    N      0               20
8     1985      25   1260      98   1133      80    H      0               18
10    1985      25   1307     103   1288      71    H      0               32


**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [36]:
index_of_max_score_difference = filtered_df['ScoreDifference'].idxmax()

In [37]:
season_highest_difference = filtered_df.loc[index_of_max_score_difference, 'Season']
winning_team_highest_difference = filtered_df.loc[index_of_max_score_difference, 'Wteam']
losing_team_highest_difference = filtered_df.loc[index_of_max_score_difference, 'Lteam']

In [38]:
print("Season, Winning Team, and Losing Team for the Game with the Highest Score Difference:")
print("Season:", season_highest_difference)
print("Wteam:", winning_team_highest_difference)
print("Lteam:", losing_team_highest_difference)

Season, Winning Team, and Losing Team for the Game with the Highest Score Difference:
Season: 1996
Wteam: 1409
Lteam: 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [39]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [40]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       N         3
       A         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [41]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [42]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [43]:
df.groupby('Season')['Wscore'].mean()

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [44]:
df.groupby('Wteam')['Wscore'].max()

Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [45]:
grouped_df = df.groupby(['Season', 'Wteam'])

In [46]:
average_winning_scores = grouped_df['Wscore'].mean()

In [47]:
team_with_highest_avg_score_by_season = average_winning_scores.groupby('Season').idxmax()

In [48]:
result_df =average_winning_scores.loc[team_with_highest_avg_score_by_season]

In [49]:
print(result_df)

Season  Wteam
1985    1328      92.800000
1986    1109      91.200000
1987    1380      95.875000
1988    1258     111.750000
1989    1258     117.315789
1990    1258     126.347826
1991    1380     112.312500
1992    1380      99.642857
1993    1380     101.875000
1994    1380     106.583333
1995    1206     102.833333
1996    1206     103.600000
1997    1254      94.142857
1998    1395     101.153846
1999    1317      95.750000
2000    1395      95.647059
2001    1395      99.315789
2002    1242      92.888889
2003    1395      88.222222
2004    1194      92.142857
2005    1311      91.333333
2006    1190      89.533333
2007    1440      99.400000
2008    1440      94.500000
2009    1377      96.200000
2010    1440      95.000000
2011    1449      90.130435
2012    1370      87.400000
2013    1377      86.500000
2014    1322      92.600000
2015    1322      90.733333
2016    1146      97.142857
Name: Wscore, dtype: float64


**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [50]:
df.groupby(['Season', 'Wteam']).count()

Daynum  Wscore  Lteam  Lscore  Wloc  Numot  ScoreDifference
Season Wteam                                                             
1985   1102        5       5      5       5     5      5                5
       1103        9       9      9       9     9      9                9
       1104       21      21     21      21    21     21               21
       1106       10      10     10      10    10     10               10
       1108       19      19     19      19    19     19               19
...              ...     ...    ...     ...   ...    ...              ...
2016   1460       20      20     20      20    20     20               20
       1461       12      12     12      12    12     12               12
       1462       27      27     27      27    27     27               27
       1463       21      21     21      21    21     21               21
       1464        9       9      9       9     9      9                9

[10172 rows x 7 columns]

**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [51]:
wins_count_df = df.groupby(['Season', 'Wteam']).size().reset_index(name='wins_count')

In [52]:
team_with_most_wins_by_season = wins_count_df.loc[wins_count_df.groupby('Season')['wins_count'].idxmax()]

In [53]:
print(team_with_most_wins_by_season)

      Season  Wteam  wins_count
217     1985   1385          27
342     1986   1181          32
818     1987   1424          33
861     1988   1112          31
1323    1989   1328          28
1551    1990   1247          29
1739    1991   1116          30
2088    1992   1181          28
2423    1993   1231          28
2665    1994   1163          27
2928    1995   1116          26
3357    1996   1269          31
3638    1997   1242          31
3943    1998   1242          33
4201    1999   1181          32
4709    2000   1409          29
4829    2001   1181          29
5121    2002   1153          30
5456    2003   1166          29
5981    2004   1390          29
6165    2005   1228          32
6452    2006   1181          30
6841    2007   1242          30
7209    2008   1272          33
7552    2009   1272          31
7870    2010   1242          32
8217    2011   1242          32
8565    2012   1246          32
8878    2013   1211          30
9462    2014   1455          33
9609    

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [54]:
average_losing_score_by_team = df.groupby('Lteam')['Lscore'].mean()

In [55]:
print("Average Losing Score by Team:")
print(average_losing_score_by_team)

Average Losing Score by Team:
Lteam
1101    60.586207
1102    59.201507
1103    64.117347
1104    64.374317
1105    61.675373
          ...    
1460    63.082915
1461    64.661972
1462    68.216117
1463    60.829213
1464    65.187063
Name: Lscore, Length: 364, dtype: float64


In [56]:
average_winning_score_by_team = df.groupby('Wteam')['Wscore'].mean()

In [57]:
print("Average Winning Score by Team:")
print(average_winning_score_by_team)

Average Winning Score by Team:
Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
          ...    
1460    75.531469
1461    75.170082
1462    79.906021
1463    71.720102
1464    73.926056
Name: Wscore, Length: 364, dtype: float64


In [58]:
teams_with_higher_average_losing_score = average_losing_score_by_team[average_losing_score_by_team > average_winning_score_by_team]

In [59]:
if not teams_with_higher_average_losing_score.empty:
    print("\nTeams with Higher Average Losing Score than Winning Score:")
    print(teams_with_higher_average_losing_score)
else:
    print("\nNo teams have a higher average losing score than winning score.")


No teams have a higher average losing score than winning score.


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [60]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [61]:
df['HighScoringGame'] = df['Wscore'].apply(lambda x: 'Yes' if x > 100 else 'No')

In [62]:
print(df.head())

   Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   
3    1985      25   1165      70   1432      54    H      0               16   
4    1985      25   1192      86   1447      74    H      0               12   

  HighScoringGame  
0              No  
1              No  
2              No  
3              No  
4              No  


**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [63]:
games_played_by_team = {}

In [64]:
for index, row in df.iterrows():
    Wteam = row['Wteam']
    Lteam = row['Lteam']

In [67]:
if Wteam in games_played_by_team:
        games_played_by_team[Wteam] += 1
else:
        games_played_by_team[Wteam] = 1

In [68]:
if Lteam in games_played_by_team:
        games_played_by_team[Lteam] += 1
else:
        games_played_by_team[Lteam] = 1

In [69]:
for team, total_games in games_played_by_team.items():
    print(f"{team}: {total_games} games played")

1386: 1 games played
1433: 1 games played


**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [81]:
highest_score_difference_by_season = {}

In [88]:
for index, row in df.iterrows():
    season = row['Season']
    winning_score = row['Wscore']
    losing_score = row['Lscore']

In [89]:
score_difference = winning_score - losing_score

In [91]:
if season in highest_score_difference_by_season:
        current_highest_score_difference = highest_score_difference_by_season[season]
        if score_difference > current_highest_score_difference['score_difference']:
            highest_score_difference_by_season[season] = {'game': index,'score_difference': score_difference}
else:
        highest_score_difference_by_season[season] = {'game': index,'score_difference': score_difference}

In [92]:
for season, data in highest_score_difference_by_season.items():
    game_index = data['game']
    score_difference = data['score_difference']
    print(f"Season {season}: Game {game_index} has the highest score difference of {score_difference}")

Season 2016: Game 145288 has the highest score difference of 13


Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [93]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [95]:
games_won = df['Wteam'].value_counts()

In [96]:
games_lost = df['Lteam'].value_counts()

In [97]:
total_games_played_by_team = games_won.add(games_lost, fill_value=0)

In [98]:
print("Total Number of Games Played by Each Team:")
print(total_games_played_by_team)

Total Number of Games Played by Each Team:
1101     76
1102    840
1103    910
1104    975
1105    447
       ... 
1460    827
1461    914
1462    954
1463    838
1464    856
Name: count, Length: 364, dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [99]:
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

In [101]:
idx_of_max_score_difference_by_season = df.groupby('Season')['ScoreDifference'].idxmax()

In [102]:
result_df = df.loc[idx_of_max_score_difference_by_season]
print(result_df)

        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
236       1985      33   1361     128   1288      68    H      0   
4731      1986      60   1314     129   1264      45    N      0   
8240      1987      51   1155     112   1118      39    H      0   
12046     1988      40   1328     152   1147      84    H      0   
16677     1989      64   1242     115   1135      45    H      0   
19502     1990      26   1181     130   1217      54    H      0   
25161     1991      73   1163     115   1148      47    H      0   
27997     1992      30   1116     128   1126      46    H      0   
33858     1993      86   1328     146   1197      65    H      0   
36404     1994      47   1228     121   1152      52    H      0   
39858     1995      26   1246     124   1404      50    H      0   
44653     1996      48   1409     141   1341      50    H      0   
49033     1997      56   1278     114   1106      34    H      0   
52600     1998      33   1395     153   1410    

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [ ]:
df[['Wscore', 'Lscore']].head()

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [ ]:
df.loc[:, ['Wscore', 'Lscore']].head()

Note the difference is the return types when you use brackets and when you use double brackets. 

In [ ]:
type(df['Wscore'])

In [ ]:
type(df[['Wscore']])

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [ ]:
df[0:3]

Here's an equivalent using iloc

In [ ]:
df.iloc[0:3,:]

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [ ]:
df.isnull().sum()

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y